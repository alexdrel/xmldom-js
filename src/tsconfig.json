{
  "compilerOptions": {
    "target": "es5",
    "module": "es6",
    "strict": true,
    "esModuleInterop": true,
    "outDir": "../dist",
    "declaration": true
  }
}
