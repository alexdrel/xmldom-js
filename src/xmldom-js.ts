export const string = (a: string) => a;

type Parser = (s: string) => any;
class ContentNode {
  constructor(public parser: Parser) { }
}
export function content(p: Parser = string) {
  return new ContentNode(p);
}
class AttributeNode {
  constructor(public parser: Parser) { }
}
export function attribute(p: Parser = string) {
  return new AttributeNode(p);
}

export type NameResolver = (jn: string, e: Element | Attr) => boolean;

export const ignoreNamespace: NameResolver = (jn, e) => e.localName == jn;

export const prefixNamespace: NameResolver = (jn, e) => e.nodeName == jn;

export function namespaces(def: string, nsmap: { [prefix: string]: string }): NameResolver {
  return (jn, e) => {
    let [p, n] = jn.split(":");
    if (n == undefined) {
      return e.localName == jn && e.namespaceURI == (e.nodeType !== e.ATTRIBUTE_NODE ? def : null);
    }
    return e.localName == n && e.namespaceURI == nsmap[p];
  }
}

function findMap<T extends Element | Attr>(list: ArrayLike<T>, name: string, ns: NameResolver, func: (e: T) => void, once = true) {
  for (let i = 0; i < list.length; i++) {
    if (ns(name, list[i])) {
      func(list[i]);
      if (once) {
        return;
      }
    }
  }
}

export function readXML(xml: Element | Document, schema: any, ns: NameResolver = ignoreNamespace): any {
  const json: any = {};
  const children = xml.children || /* IE */ (xml.childNodes as any as HTMLCollection);
  Object.keys(schema).forEach(k => {
    const s = schema[k];
    let v: any;
    if (s instanceof ContentNode) {
      if (k === "") {
        v = s.parser(xml.textContent as string);
      } else {
        findMap(children, k, ns, e => v = s.parser(e.textContent as string));
      }
    } else if (s instanceof AttributeNode && xml instanceof Element) {
      findMap(xml.attributes, k, ns, e => v = s.parser(e.textContent as string));
    } else if (Array.isArray(s)) {
      v = [];
      findMap(children, k, ns, e => v.push(readXML(e, s[0], ns)), false);
    } else /* Structured element */ {
      findMap(children, k, ns, e => v = readXML(e, s, ns));
    }
    if (v !== undefined) {
      json[k] = v;
    }
  });
  return json;
}
