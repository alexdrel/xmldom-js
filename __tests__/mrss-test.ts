import { content, string, readXML, attribute, prefixNamespace } from "../src/xmldom-js";


const rssString = `<?xml version='1.0' encoding='UTF-8'?>
<rss xmlns:atom="http://www.w3.org/2005/Atom" xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:media="http://search.yahoo.com/mrss/" version="2.0">
  <channel>
	<link>http://example.com</link>
	<description>MRSS</description>
  <title>MRSS Title</title>

	<item>
		<title>Item 1</title>
		<media:title>Item 1</media:title>
		<media:content url="http://example.com/1.mp4" medium="video"/>
		<media:thumbnail url="http://example.com/1.jpg" height="360" width="640"/>
		<media:comment>Item 1 Comment</media:comment>
	</item>

  <item>
		<title>Item 2</title>
		<media:title>Item 2</media:title>
		<media:content url="http://example.com/2.mp4" medium="video"/>
		<media:thumbnail url="http://example.com/2.jpg" height="360" width="640"/>
		<media:comment>Item 2 Comment</media:comment>
	</item>

  <item>
		<title>Item 3</title>
		<media:title>Item 3</media:title>
		<media:content url="http://example.com/3.mp4" medium="video"/>
		<media:thumbnail url="http://example.com/3.jpg" height="360" width="640"/>
		<media:comment>Item 3 Comment</media:comment>
  </item>
  </channel>
</rss>
`;

const rssXml = new DOMParser().parseFromString(rssString, "text/xml");
const schema = {
  rss: {
    channel: {
      title: content(string),
      item: [{
        title: content(string),
        "media:content": { url: attribute() },
        "media:thumbnail": { url: attribute(string) },
        "media:nope": { url: attribute(string) },
        "media:comment": content(string)
      }]
    }
  }
};


describe("rss", () => {
  const rssJson = readXML(rssXml, schema, prefixNamespace);

  test("Check the document", () => {
    expect(rssJson.rss).toBeDefined();
    expect(rssJson.rss.channel).toBeDefined();
    expect(rssJson.rss.channel.item.length).toBeDefined();
  });
  test("Check the document shot", () => {
    expect(rssJson.rss.channel).toMatchSnapshot();
  });
  test("Check items", () => {
    expect(rssJson.rss.channel.item.length).toBe(3);
  });
});
