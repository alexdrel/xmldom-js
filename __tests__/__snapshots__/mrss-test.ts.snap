// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rss Check the document shot 1`] = `
Object {
  "item": Array [
    Object {
      "media:comment": "Item 1 Comment",
      "media:content": Object {
        "url": "http://example.com/1.mp4",
      },
      "media:thumbnail": Object {
        "url": "http://example.com/1.jpg",
      },
      "title": "Item 1",
    },
    Object {
      "media:comment": "Item 2 Comment",
      "media:content": Object {
        "url": "http://example.com/2.mp4",
      },
      "media:thumbnail": Object {
        "url": "http://example.com/2.jpg",
      },
      "title": "Item 2",
    },
    Object {
      "media:comment": "Item 3 Comment",
      "media:content": Object {
        "url": "http://example.com/3.mp4",
      },
      "media:thumbnail": Object {
        "url": "http://example.com/3.jpg",
      },
      "title": "Item 3",
    },
  ],
  "title": "MRSS Title",
}
`;
